cli/pkg/edgeclient/client.go

```diff
@@ -96,13 +102,370 @@ func (c *Client) AddNewModelWithImage(
+	modelVersion string,
+	namespace string,
+	kubeConfig string,
+	parameters map[string]interface{},
+) (*PipelineRun, error) {
+	var s3SecretName string
+	var testDataConfigMapName string
+	var addDirWorkspace bool
+
+	if s3s, ok := parameters["s3SecretName"]; !ok {
+		return nil, fmt.Errorf("s3SecretName pipeline parameter is required")
+	} else if s3s, ok := s3s.(string); !ok {
+		return nil, fmt.Errorf("s3SecretName pipeline parameter must be a string")
+	} else {
+		s3SecretName = s3s
+	}
+
+	if tdc, ok := parameters["testDataConfigMapName"]; !ok {
+		return nil, fmt.Errorf("testDataConfigMapName pipeline parameter is required")
+	} else if tdc, ok := tdc.(string); !ok {
+		return nil, fmt.Errorf("testDataConfigMapName pipeline parameter must be a string")
+	} else {
+		testDataConfigMapName = tdc
+	}
+
+	if adw, ok := parameters["addDirWorkspace"]; !ok {
+		return nil, fmt.Errorf("addDirWorkspace pipeline parameter is required")
+	} else if adw, ok := adw.(string); !ok {
+		return nil, fmt.Errorf("addDirWorkspace pipeline parameter must be a boolean")
+	} else {
+		addDirWorkspace = adw == "true"
+	}
+
+	params, err := toTektonParams(modelName, modelVersion, parameters)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert parameters to tekton params: %w", err)
+	}
+
+	pipelineRun := &tektonv1.PipelineRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:    namespace,
+			GenerateName: "aiedge-e2e-",
+			Labels: map[string]string{
+				"tekton.dev/pipeline": "aiedge-e2e",
+			},
+		},
+		Spec: tektonv1.PipelineRunSpec{
+			TaskRunTemplate: tektonv1.PipelineTaskRunTemplate{
+				ServiceAccountName: "pipeline",
+			},
+			Params: params,
+			PipelineRef: &tektonv1.PipelineRef{
+				Name: "aiedge-e2e",
+			},
+			Workspaces: []tektonv1.WorkspaceBinding{
+				{
+					Name: "build-workspace-pv",
+					VolumeClaimTemplate: &corev1.PersistentVolumeClaim{
+						Spec: corev1.PersistentVolumeClaimSpec{
+							AccessModes: []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
+							Resources: corev1.VolumeResourceRequirements{
+								Requests: corev1.ResourceList{
+									corev1.ResourceStorage: resource.MustParse("1Gi"),
+								},
+							},
+						},
+					},
+				},
+				{
+					Name: "s3-secret",
+					Secret: &corev1.SecretVolumeSource{
+						SecretName: s3SecretName,
+					},
+				},
+				{
+					Name: "test-data",
+					ConfigMap: &corev1.ConfigMapVolumeSource{
+						LocalObjectReference: corev1.LocalObjectReference{
+							Name: testDataConfigMapName,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	if addDirWorkspace {
+		pipelineRun.Spec.Workspaces = append(
+			pipelineRun.Spec.Workspaces,
+			tektonv1.WorkspaceBinding{
+				Name:     "workspace",
+				EmptyDir: &corev1.EmptyDirVolumeSource{},
+			},
+		)
+	}
+
+	config, _ := clientcmd.BuildConfigFromFlags("", kubeConfig)
+	tektonClient, _ := tektonclientset.NewForConfig(config)
+
+	createdPipelineRun, err := tektonClient.TektonV1().PipelineRuns(namespace).Create(
```

Comment:
Having the earlier "calculation" of constructing the PipelineRun object in a separate function than this "action" of sending it to the server might be a useful way to break it down a little? It should make testing of the different conditions in the construction of the object easier to test, without needing to worry about the server, for example.

Reply:
Yeah makes sense, refactore.